Edited azure-container-apps/README.md for style and consistency (#290)
1Password/scim-examples: azure-container-apps/ReadMe.md

| Volume    | Number of users | CPU   | memory |
| --------- | --------------- | ----- | ------ |
| Default   | <1,000          | 0.25  | 0.5Gi  |
| High      | 1,000–5,000     | 0.5   | 1.0Gi  |
| Very high | >5,000          | 1.0   | 1.0Gi  |

If you're provisioning more than 1,000 users, update the resources assigned to the SCIM bridge container to follow these recommendations. The resources specified for the Redis container do not need to be adjusted.

### Default

To revert to the default specification that is suitable for provisioning up to 1,000 users, run the following command:

```sh
az containerapp update -n $ConAppName -g $ResourceGroup --container-name op-scim-bridge \
  --cpu 0.25 --memory 0.5Gi
```

### High volume deployment

If you're provisioning up to 5,000 users:

```sh
az containerapp update -n $ConAppName -g $ResourceGroup --container-name op-scim-bridge \
  --cpu 0.5 --memory 1.0Gi
```

### Very high volume

If you're provisioning more than 5,000 users:

```sh
az containerapp update -n $ConAppName -g $ResourceGroup --container-name op-scim-bridge \
  --cpu 1.0 --memory 1.0Gi
```

> ## Get help

Reviewer: Another option instead of `Get Help` could be `Common Troubleshooting Tips`.

Author: I think "Get help" is ideal since it's consistent with other articles, and some of the headers aren't _technically_ troubleshooting, but rather specs. We could rename the headers though...

# Deploy 1Password SCIM Bridge on Azure Container Apps using the Azure Portal

_Learn how to deploy 1Password SCIM Bridge on the [Azure Container Apps](https://azure.microsoft.com/en-us/products/container-apps/#overview) service._

This deployment consists of two [containers](https://learn.microsoft.com/en-us/azure/container-apps/containers): One for the SCIM bridge and another for Redis. There's also an [ingress](https://learn.microsoft.com/en-us/azure/container-apps/ingress-overview) for the SCIM bridge container. There are a few benefits to deploying 1Password SCIM Bridge on Azure Container Apps:

- For standard deployments, the service will host your SCIM bridge for ~$16 USD/month (as of January 2024). Container Apps pricing is variable based on activity, and you can learn more on [Microsoft's pricing page](https://azure.microsoft.com/en-us/pricing/details/container-apps/).
- You don't need to manage a DNS record. Azure Container Apps automatically provides a unique one for your SCIM bridge domain.
- Azure Container Apps automatically handles TLS certificate management on your behalf.

> - The SCIM bridge can be deployed directly to Azure from the Portal or via the command line in your local terminal. There is no requirement to clone this repository.

Reviewer: Small suggestion to highlight that this guide is for the portal deployment and have a link out to the AKB guide for Shell based deployments. Feel free to adjust the wording, as there is something that is not 100% sitting correctly with my suggestion. 

```suggestion
- The SCIM bridge can be deployed directly to Azure from the Portal using this guide or via the Azure Shell or command line tools in your local terminal following our [support guide](https://support.1password.com/scim-deploy-azure/). There is no requirement to clone this repository.
```

**Table of contents:**

- [Before you begin](#before-you-begin)
- [Step 1: Configure the `scimsession` credentials](#step-1-configure-the-scimsession-credentials)
- [Step 2: Create the container app](#step-2-create-the-container-app)
- [Step 3: Configure and deploy your SCIM bridge](#step-3-configure-and-deploy-your-scim-bridge)
- [Step 4: Connect your identity provider](#step-4-connect-your-identity-provider)
- [Update your SCIM Bridge](#update-your-scim-bridge)
- [Appendix: Resource recommendations](#appendix-resource-recommendations)
- [Appendix: Customize your deployment](#appendix-customize-your-deployment)
- [Get help](#get-help)

## Before you begin

Before you begin, familiarize yourself with [PREPARATION.md](/PREPARATION.md) and complete the necessary steps there. You'll also need the following:

- An Azure account with permissions to create a Container App.

> [!NOTE]
> If you don't have a Azure account, you can sign up for a free trial with starting credit: https://azure.microsoft.com/en-us/free/

> - [Azure CLI tool](https://learn.microsoft.com/en-us/cli/azure/install-azure-cli) with the ContainerApp Extension or access to the [Azure Cloud Shell](https://learn.microsoft.com/en-us/azure/cloud-shell/quickstart) if you'd like to deploy using the command line.

Reviewer: I guess this is not really a true requirement here, since we are guiding how to deploy with the portal GUI. 
```suggestion
```

### In this folder

> - [`aca-op-scim-bridge.yaml`](./aca-op-scim-bridge.yaml): an Container App configuration file of a container app for the 1Password SCIM bridge.

Reviewer: I guess the description of this file should reference that this .yaml is used solely for the https://support.1password.com/scim-deploy-azure/ deployment method.

Author: Ah good point!

```suggestion
- [`aca-op-scim-bridge.yaml`](./aca-op-scim-bridge.yaml): a Container App configuration file for 1Password SCIM Bridge that you can use when you deploy via the command line.
```

	4. **CPU and Memory**: Choose **0.25 CPU cores, 0.5 Gi memory**, which should be the default option.
	5. **Environment variables**: Create two with the following details:
		1. `OP_SESSION` as the Name and  `""` as the Value.
		2. `OP_REDIS_URL` as the Name and `redis://localhost:6379` as the Value.

3. Click **Next: Bindings >** at the bottom of the page, then click **Next : Ingress >** to continue.
4. Click **Enabled** to turn on ingress, then adjust the following:
	1. **Ingress Traffic**: Choose **Accept traffic from anywhere**.
	2. **Target Port**: Enter `3002`.

5. Click **Review + create** at the bottom of the page, then wait a moment for the page to load and click **Create.**

After the deployment is complete, click **Go to resource**, then continue to step 3.

> [!NOTE]
> The deployment will fail because it's expecting the second container to deploy successfully. Continue to step 3 to fully deploy the SCIM bridge.

## Step 3: Configure and deploy your SCIM bridge

### 3.1: Create a secret with your `scimsession` file

1. Choose **Secrets** from the Settings section in the sidebar.
2. Click **Add**, then enter `scimsession` in the Key field.

> 3. Paste the base64 value into the **Value** field. This is the `scimsession` item that you added to your account in [step 1](#step-1-configure-the-scimsession-credentials).

Reviewer: Small wording change to highlight we need the base64 value of the scimsession item from step 1. 
```suggestion
3. Paste the base64 value into the **Value** field. This is the `Base64 value` in your `scimsession` item that you added to your account in [step 1](#step-1-configure-the-scimsession-credentials).
```

Reviewer 2: I'd not use ` `` ` here to surround `Base64 value`, as that isn't "code". Italics / bold make more sense here in my opinion.

<details>
<summary>Connect Google Workspace using the Azure Portal</summary>

#### Step 1: Get your Google service account key

1. Follow the steps to [create a Google service account, key, and API client](https://support.1password.com/scim-google-workspace/#step-1-create-a-google-service-account-key-and-api-client).

2. Copy and paste the following command for your shell to create a base64 value for your key. Change `<keyfile>` to the filename generated by Google for your service account.
    - **Bash**:
     ```bash
     cat /path/to/<keyfile>.json | base64
     ```
    - **PowerShell:**
     ```pwsh
     [Convert]::ToBase64String([IO.File]::ReadAllBytes((Join-Path $PWD.Path '/path/to/<keyfile>.json')))
     ```

3. Open the Azure Portal and go to the [Container Apps](https://portal.azure.com/#view/HubsExtension/BrowseResource/resourceType/Microsoft.App%2FcontainerApps) page.

4. Choose **Secrets** from the Settings section in the sidebar.

5. Click **Add**, the fill out the following fields:
    1. **Key**: Enter `workspace-credentials`.
    2. **Value**: Paste the base64 value you just created.

6. Click **Add**, then wait until the secret is created.

> 7. Add two environment variables to the new secret:
>     1. `OP_WORKSPACE_CREDENTIALS`
>     2. `OP_WORKSPACE_SETTING`
> 

Reviewer: Remove this step. This is out of place and was missed on previous reviews. Line 184 has the user save the workspace-credentials secret, then we go on in Line 214 to add the workspace-settings secret. We are adding both environment variables in line 218-226 as part of Step 3: Add the secrets to your container

```suggestion
```

     2. **Value**: Enter the base64 value you just created.

3. Click **Add**, then wait until the secret is created.
4. Choose **Containers** from Application section in the sidebar.
5. Click **Edit and deploy**.
6. Hover over your **op-scim-bridge** container and select it, then click **Edit**.
7. Scroll down to the environment variables and add a new one. Then fill out the following fields:
      1. **Name**: Enter `OP_WORKSPACE_CREDENTIALS`.
      2. **Source**: Choose **Reference a secret**.
      3. **Value**: Select the `workspace-credentials` secret.

8. Create another new variable and fill out the following fields:
      1. Name: **OP_WORKSPACE_SETTINGS**
      2. **Source**: Choose **Reference a secret**.
      3. **Value**: Select the `workspace-settings` secret.

9. Click **Save**, then click **Create**.

</details>

<hr>

## Update your SCIM bridge

> [!TIP]
>
> Check for 1Password SCIM Bridge updates on the [SCIM bridge release page](https://app-updates.agilebits.com/product_history/SCIM).

You can update your SCIM bridge in the [Azure Cloud Shell](#in-the-azure-cloud-shell-or-az-cli) or the [Azure Portal](#in-the-azure-portal).

> ### In the Azure Cloud Shell or AZ CLI

Reviewer: Do we remove this section about updating in the Azure Cloud Shell or AZ CLI and point to our https://support.1password.com/scim-update/#azure-container-apps guide?

Author: Yes! I'll do that.

| Volume    | Number of users | CPU   | memory |
| --------- | --------------- | ----- | ------ |
| Default   | <1,000          | 0.25  | 0.5Gi  |
| High      | 1,000–5,000     | 0.5   | 1.0Gi  |
| Very high | >5,000          | 1.0   | 1.0Gi  |

If you're provisioning more than 1,000 users, update the resources assigned to the SCIM bridge container to follow these recommendations. The resources specified for the Redis container do not need to be adjusted.

### Default

To revert to the default specification that is suitable for provisioning up to 1,000 users, run the following command:

```sh
az containerapp update -n $ConAppName -g $ResourceGroup --container-name op-scim-bridge \
  --cpu 0.25 --memory 0.5Gi
```

### High volume deployment

If you're provisioning up to 5,000 users:

```sh
az containerapp update -n $ConAppName -g $ResourceGroup --container-name op-scim-bridge \
  --cpu 0.5 --memory 1.0Gi
```

### Very high volume

If you're provisioning more than 5,000 users:

```sh
az containerapp update -n $ConAppName -g $ResourceGroup --container-name op-scim-bridge \
  --cpu 1.0 --memory 1.0Gi
```

> 

Reviewer: I am not 100% sold on removing the Azure Cloud Shell method for connecting google workspace completely. I feel we need to keep this info somewhere. Perhaps as additional appendix related info. We do call out how to upgrade resources in the appendix, but only show the cloud shell method to do so (which I am fine with).

Reviewer: @Penderworth I feel that from our team, we are trying to ensure we are documenting in GH the ability to scale the SCIM bridge to meet higher demand requirements (if required be the customer). While this may not be the best place in the GUI readme, I think some items such as resources requirements and get help (troubleshooting tips) might be great in another md file with this info. That way it is available for both the GUI deployment method and the Cloud shell method in AKB.

Author: That's a good point. I'll add it back and update it to match the rest of the style here!

Author: @black-bryan Alright these are back in https://github.com/1Password/scim-examples/pull/290/commits/ac46bcbd76545ce42c5c8242722d5b47b6bff64b. Let me know how they look.

5. Update your identity provider configuration with the new bearer token.
    </details>

<details>
<summary>Replace your `scimsession` secret using the Azure Portal</summary>

1. Run the following command for your shell to get the Base64 encoded contents of your new `scimsession` file.

    - **Bash**:

        ```bash
        cat ./scimsession | base64
        ```

    - **PowerShell**:

        ```pwsh
        [Convert]::ToBase64String([IO.File]::ReadAllBytes((Join-Path $PWD.Path 'scimsession')))
        ```

2. Copy the output value to your clipboard to use it in the deployment.
3. Open the Azure Portal and go to the [Container Apps](https://portal.azure.com/#view/HubsExtension/BrowseResource/resourceType/Microsoft.App%2FcontainerApps) page.
4. Choose **Secrets** from the Settings section in the sidebar.
5. Edit the `scimsession` secret and paste the new value of your secret.
6. Select the checkbox and click **Save**.
7. Choose the **Revisions** from the Application section in the sidebar.
8. Click your current active revision and choose **Restart** in the details pane.
9. Open your SCIM bridge URL in a browser and enter your bearer token to test the bridge.

> 10. Update your identity provider configuration with the new bearer token.

Author: I left this as is, but am not really sure what it means. They aren't using their bearer token to connect their IdP...

Reviewer: I am confused by this comment. When the credentials are regenerated in 1Password, both the scimsession secret on the deployment and their provisioning connection from their IdPs (except GW) require the enterprise/business application be updated with the new bearer token to continue functioning with the new scimsession.
